src/main/webapp/filter_tag_area.js

```diff
@@ -0,0 +1,143 @@
+    this.filterTypeInput.setAttribute("class", "filter-input-area");
+    this.filterTypeInput.setAttribute("placeholder", "Filter by:");
+    this.filterTypeInput.addEventListener('keypress', (e) => {
+      if (e.key === 'Enter') {
+        this.filterParamLabel.textContent = `${this.filterTypeInput.value}:`;
+        this.filterTypeInput.removeChild(this.filterDataList);
+        this.filterEntryArea.removeChild(this.filterTypeInput);
+        this.filterEntryArea.appendChild(this.filterParamInput);
+        this.filterEntryArea.appendChild(this.filterParamLabel);
+      }
+    });
+
+    this.filterDataList = document.createElement("datalist");
+    this.filterDataList.setAttribute("id", "filter-datalist");
+    this.optionMap = new Map();
+    this.optionMap.set("Organization Name", "orgNames");
+    this.optionMap.set("Address", "orgStreetAddresses");
+    this.optionMap.set("Available Resources", "resourceCategories");
+    for (const optionKey of this.optionMap.keys()) {
+      const option = document.createElement("option");
+      option.value = optionKey;
+      this.filterDataList.appendChild(option);
+    }
+
+    /* After the type has been chosen, the actual filter param is entered here */
+    this.filterParamInput = document.createElement("input");
+    this.filterParamInput.setAttribute("class", "filter-input-area");
+    this.filterParamLabel = document.createElement("div");
+    this.filterParamLabel.setAttribute("class", "filter-tag-label");
+    this.filterParamInput.addEventListener('keypress', (e) => {
+      if (e.key === 'Enter') {
+        /* If the user enters an unsupported type, it is ignored on the servlet side */
+        this.parentSearchArea.setUrlParamValue(this.optionMap.get(this.filterTypeInput.value), this.filterParamInput.value);
+        this.filterParamInput.value = "";
+        this.filterTypeInput.value = "";
+        this.filterEntryArea.removeChild(this.filterParamInput);
+        this.filterEntryArea.removeChild(this.filterParamLabel);
+        this.activeFilterArea.removeChild(this.filterEntryArea);
+      }
+    });
+
+    /* This button does not move- when clicked it opens a filter entry area */
+    this.addFilterButton = document.createElement("div");
+    this.addFilterButton.setAttribute("class", "filter-tag-area");
+    this.addFilterButton.setAttribute("id", "add-filter-button");
+    this.addFilterButton.textContent = "+ Add Filter";
+    this.addFilterButton.addEventListener('click', () => {
+      /* If the user is on the second part of entering a filter, add filter is disabled */
+      if (!this.filterEntryArea.contains(this.filterParamInput)) {
+        this.filterTypeInput.appendChild(this.filterDataList);
+        this.filterEntryArea.appendChild(this.filterEntryClose);
+        this.filterEntryArea.appendChild(this.filterTypeInput);
+        this.activeFilterArea.appendChild(this.filterEntryArea);
+      }
+    });
+    this.activeFilterArea.appendChild(this.addFilterButton);
```

Comment:
Is this class meant to be the area that holds all the filters, a single filter, or both? Would the code be simpler if you decomposed it into the 2 separate concepts?

Reply:
This class holds the add filter button, the tag that pops up when you click add filter, and then whatever filter is being applied. Considering how connected these concepts are, I think it's definitely simplest to encapsulate them all into the same class.

Reply:
I'm not sure I agree that it makes it simpler! Encapsulating code (even if there's more of it) results in less that needs to be understood, as things get pulled into helper functions, documented, and decoupled. 

I'd go so far as to say that teasing apart where you can break apart related concepts is one of the most powerful things you can do to simplify a design.

Reply:
The FilterTagArea class now holds the add filter button / text box, and the tags themselves are represented by the FilterTag class and are added to the activeFilterArea of this class.

Reply:
If you had a library with the following interface:
```
class FilterEntry {
  // Constructs a new filter entry row.
  // Starts with:
  //   - a text area to enter the filter field
  //   - hasFilterField == false, hasFilterValue == false
  // After the filter field is entered:
  //   - there is a second text area to enter the filter value
  //   - hasFilterField == true, hasFilterValue == false
  // After the filter value is entered:
  //   - hasFilterField == true, hasFilterValue == true
  // There is also a close button in the entry row. When clicked
  // the filter row will be removed, and onRemoveCallback(this) is called
  constructor(div parentElement, onRemoveCallback)

  bool hasFilterField()
  string getFilterField()

  bool hasFilterValue()
  string getFilterValue()
}
```

Would it make the implementation of FilterTagArea easier?